### ISSUE TITLE
Avoiding oversimplication

### ISSUE THREAD

**Issue by wwitzel** [PyProveIt/Prove-It#251]
There are a number of @prover methods that instantiate a theorem to produce an implication and the derive the consequent.  Or, similary, produce an equality and then derive the left or right side.  If there is auto-simplification and this method is executed again, you will end up deriving "|- T" which is useless.  One fix is to use "auto_simplify=False" (of "preserve_all=True" which may be overkill).  However, there are times when simplification is desired, but not this oversimplification.  Is there a way to automatically know the difference and avoid oversimplifying to True when it isn't useful?  Or, failing that, what is the easiest way to indicate the desired behavior?

**Comment by wwitzel** [comment]
It's easier to handle the implication case.  If you have "A => B" at the top level (e.g., "|- A => B"), don't every simplify B to True in the same way that you never simplify all the way to "|- T" and for the same reason.  "A => T" is always True -- it is a vacuous statement, so don't simplify it to this point at the top level.

**Comment by wwitzel** [comment]
For the "derive_left_via_equality" and "derive_right_via_equality", it may be best to handle this on a case-by-case basis.  It could be handled using "preserve_all=True" or "auto_simplify=False" or tailored more specifically to the situation.

**Comment by wwitzel** [comment]
Not, I'm implementing the implication-specific feature in a way that applies recursively.  For example, "A => (B => C)" at the top-level will simplify neither (B => C) nor C.  This is good.

**Comment by wwitzel** [comment]
There are cases that come up involving Equals expressions that we don't want to oversimplify: InnerExpr.simplify, and OperationOverInstances bundling/unbundling are examples.  Some "defaults" configuration might be nice for this.  But what should it be called.  no_simplify_lhs_to_true and no_simplify_to_rhs_true??

**Comment by wwitzel** [comment]
Here is a different strategy altogether that is much simpler and may be for the best.  For auto-simplification, don't use "known evaluations".  Instead, just use "shallow_simplifications" along the way.  The idea was, "what could be simpler than an evaluation if it is known?".  However, that leads to this undesirable behavior in which things are oversimplified to vacuous statements.  Maybe the use of "known evaluations" auto-simplification leads to counter-intuitive behavior that is undesirable.  Interestingly, nothing appears to break (no demonstrations page or theorem proof) when disabling this "feature".  So it isn't being used.  We could have it as an optional auto-simplification strategy, but turn it off by default.  Perhaps call it defaults.use_known_evaluations_in_auto_simplification; that is kind of wordy but may be okay.

To illustrate the point, here is a completely arbitrary example:
Let's say that we are trying to prove there are more apples (Na) than bananas (Nb).  Incidentally, we have learned that Na=10.  We instantiate some theorem to derive something like:
 [(Na + 1) - (Nb + 1) <= 0] => False
This should auto-simplify to
(Na - Nb <= 0) => False
From here we can prove "(Na < Nb) => False" and prove "Na > Nb" via contradiction.  But if we oversimplified Na to 10 along the way, that would go against our goal.
